### ISSUE TITLE
Input Approximation Information from FMU

### ISSUE THREAD

**Issue by clagms** [modelica/fmi-standard#794]
This is an adaptation of <https://github.com/modelica/fmi-standard/issues/623#issuecomment-536640784>.

I propose that FMI 3.0 standardizes the expectations of the input approximation function of the FMU regarding each individual input.

Interpolations and Extrapolations are two different function approximation schemes, and right now there's no way of distinguishing which an FMU implements.

### Example:
_____
Consider an FMU A, that implements an input **interpolation** scheme. When FMU A is at time $t$, and a setReal is called in that FMU A, the FMU A will assume that it is receiving an input that was computed by another FMU B that is already at time $t+H$.
This assumption has to be made because there is no way of knowing the timestamp of an input when a call is made to setReal.
We may argue that the timestamp of a value corresponds to the timestamp of the internal state of the FMU that computed it (although one has to be careful with this definition, and more details are given in the paper cited below).
So, a correct master algorithm will do something like this, every cosim step:
```
doStep(B, t, H)
getReal(B, y, v)
setReal(A, u, v)
doStep(A, t, H)
```
An incorrect master algorithm could do something like this, every cosim step:
```
getReal(B, y, v)
setReal(A, u, v)
doStep(B, t, H)
doStep(A, t, H)
```
_____

In the above example, the second master algorithm is incorrect because FMU A is assuming that the value it gets is from a "future" timestamp, and will consequently perform an interpolation on it.
So, FMU A only works with Gauss Seidel like master algorithms.

Without explicit timestamps on the inputs, the implementation of a interpolation is different than the implementation of an extrapolation.
The following figure is an attempt at depicting the difference on the inputs that FMU A will give to its internal numerical solver, when used correctly (top) and incorrectly (bottom):
![image](https://user-images.githubusercontent.com/871103/65947617-f454a400-e438-11e9-8d56-bf07829a829b.png)

I also attach the paper where me and the co-authors hopefully explain this matter a little bit better 😋 
[cosimcpspaper.pdf](https://github.com/modelica/fmi-standard/files/3675290/cosimcpspaper.pdf)

Some people may argue that all FMUs should implement their input approximations expecting an input at time $t$, thereby making correct the second master algorithm in the above example, and incorrect the first master algorithm.
However, note that the gauss seidel implicit algorithm, which is one of the master algorithms proposed by Kübler and Schiehlen's paper (which the FMI claims to be consistent with) is not allowed.

Related to the above is the meaning of the canInterpolateInputs flag.

## canInterpolateInputs flag

It's not clear which inputs this flag refers to.
If the FMI is going to supported nested FMUs, then this flag needs to be deleted, and a flag per scalar input needs to be introduced.
As stated in Section 4.1.2., `an FMU can be a combination of a hybrid ODE and a sampled data system.` In that case, it is not necessarily the case that all continuous inputs are being approximated by the FMU.

Ideally, if we solve the above issue, we solve this ambiguity as well.

PS - This issue has been discussed with, and approved by @CThuleHansen .

**Comment by pmai** [comment]
Maybe I'm confused here, but when you say

> So, a correct master algorithm will do something like this, every cosim step:
    doStep(B, t, H)
    getReal(B, y, v)
    setReal(A, u, v)
    doStep(A, t, H)
> An incorrect master algorithm could do something like this, every cosim step:
    getReal(B, y, v)
    setReal(A, u, v)
    doStep(B, t, H)
    doStep(A, t, H)

This seems to me to be reversed _in the context of FMI 1.0/2.0 co-simulation semantics_: These more or less actually assume the second set of steps. The former just does not seem valid for FMI co-simulation, more or less precisely for the reasons you then lay out.

In the second set of steps, what you call the timestamp of all exchanged signals is actually very clear: At time t, outputs are read and inputs set for time t, and then a co-simulation step to t + h is performed, leading to the next signal exchange happening at t+h with signals valid for t+h.

This is also made very clear in section 4.1.1 of the FMI 2.0.1 standard:

> 4. The slave is able to interrupt the simulation when tc_i is reached.
> 5. During the interrupted simulation the slave (and its individual solver) can receive values for inputs u(tc_i) and send values of outputs y(tc_i).

This works perfectly well with the second set of steps, and matches the example master algorithm in the specification.

If you want to do advanced master algorithms, it is the master algorithm that needs to ensure that the values it sets are valid for time tc_i (i.e. the current communication time). It is not able to pass other values to the FMU and hope that the FMU magically inter- or extrapolates them to get the values for tc_i (how would it do that, without knowing the master algorithm and having no information about the time they actually apply to)?

Given that this interpretation of FMI co-simulation makes moot your problem, is consistent with the explicit statements of the spec and the sample master-algorithm in the spec, and seems to be what most implementations actually assume, whereas a different interpretation would lead to lots of problems in the spec, lots of stuff not being defined, I don't see how one would arrive at a different interpretation.

And while section 4.1.2 does indeed state that

> Definition (4.1) is consistent with the definition of co-simulation by (Kübler, Schiehlen 2000).

this only applies to definition 4.1, which again clearly bases all calculations of x_i+1 and y_i+1 only on x_i and u_i, never on any other time points.

So it seems to me that you are reading too much into that sentence, and are assuming things which are not actually in the spec.

But again, maybe I'm missing something here, since it seems to me to be very clear...

**Comment by clagms** [comment]
Thank you for this @pmai !

My main concern is that the paper (Kübler, Schiehlen 2000) actually allows for both master algorithms that I proposed earlier (see fig. 6 of that paper).

Moreover, in general, it is reasonable to allow that an FMU can compute its state x_{i+1} from u_{i+1}, as long as not all FMUs in the cosim do so. In our experience, this leads to better co-simulation solutions (more accurate, energy conserving, etc.).

See, for example, the following paper, where the only way to get a good solution was to have adapt some FMUs behave in the manner consistent with the first master algorithm I proposed above:
- Gomes, Cláudio, Bentley James Oakes, Mehrdad Moradi, Alejandro Torres Gamiz, Juan Carlos Mendo, Stefan Dutre, Joachim Denil, and Hans Vangheluwe. “HintCO - Hint-Based Configuration of Co-Simulations.” In International Conference on Simulation and Modeling Methodologies, Technologies and Applications, 57–68. Prague, Czech Republic, 2019. https://doi.org/10.5220/0007830000570068.

Also see this paper, 
Arnold, Martin. “Stability of Sequential Modular Time Integration Methods for Coupled Multibody System Models.” Journal of Computational and Nonlinear Dynamics 5, no. 3 (May 14, 2010): 9. https://doi.org/10.1115/1.4001389.
where Martin describes a sequential method that is much like the first master algorithm I proposed.
See Eq 5a, where you see the first system being integrated, followed by the second, in eq 8. The second system gets the solution for the first in the interval t -> t+H.


In short: if the first master is forbidden by the standard, then please remove references to Kubler and avoid saying that the FMI standard is consistent with that definition.
I'm fine with either support or no support for sequential master algorithms, as long as it is clear where the fmi standard stands.

If you all agree that there should be no support, then I can propose some modifications to the text to hopefully avoid inducing the reader into thinking that the fmi supports this.

**Comment by chrbertsch** [comment]
I agree with the original comment in https://github.com/modelica/fmi-standard/issues/623#issuecomment-536640784 that the notion of "Input interpolation" is misleading. 
What we do in fact is an extrapolation form time t_i to t_{i+1} based on the inputs values and its derivative(s) at time t_i.
So I think the name of the flag canInterpolateInputs is misleading.

One could realize an input _interpolation_ using intermediate variable access https://fmi-standard.org/docs/3.0-dev/#state-intermediate-update-mode-co-simulation

**Comment by chrbertsch** [comment]
I propose to treat this ticket also as master ticket for #603, #622,  https://github.com/modelica/fmi-standard/issues/623#issuecomment-536640784

In Berlin in a group with @masoud-najafi, @friedrichatgc , @klausschuch we discussed this; here are my notes: 
- not specifying the input derivative order could create a problem or inefficiency regarding memory allocation in the FMU to store the derivative values (of at the beginning unknown and possibly varying order)
- the order of passed values to the `fmi3SetInputDerivatives` for multiple VRs and derivative order >1 is not clearly defined. 
- for interpolation and usage of known input values on the whole interval t_i to t_i+1 one should in FMI 3.0 use Intermediate Variable Access
- `canInterpolateInputs` is in fact an extrapolation from point t_i
- one could and possibly should specify the concrete extrapolation polynomial. Only then could the master really influence the FMU on the whole interval t_i to t_i+1 in a defined way. 
E.g. if the FMU uses 2nd input derivatives, the master must known this
- it has to be defined if the derivatives set at t_i have to be the real derivatives, or just  values that are used in the extraplation polynomial (e.g. to reach a certain value of the inputs at t_i+1)
- (today almost no-one uses input derivative order higher than 1)
- ... please extend of correct if I have forgotten something

I do not remember well what we agreed on as next steps:
@friedrichatgc :  will you create a PR for this? Shall we discuss this at the next regular design meeting or should we have a separate discussion on this?

**Comment by chrbertsch** [comment]
Online Design Meeting: 

Markus F.: As we have Intermediate Variable Access in BCS (see https://github.com/modelica/fmi-standard/issues/836), we do not have to define a polynomial how the the input derivatives can be used.

Antoine: This is completely different to the FMI 2.0 approach. Is IVA optional?

Christian B.: Yes

Antoine: do we want the only way for extrapolation to use IVA?

Markus: There are two ways: Either use IVA, or in as in FMI 2.0 by setting input derivatives. But in the second case we must clarify what to do with this information (e.g. a specific taylor polynomial).
Proposal: Define theat the input derivative is really the mathematical derivative. Then it is up to the FMU what to do with this.

Antoine: there was a specification in FMI 1.0 Appendix D. But not part of FMI 2.0

Markus: was it normative? This is unclear.

Antoine: this was by Martin Arnold.

Markus: the aim was to provide an exact polynomial. If this is the intention, why allow the FMU to do what it wants? But now we have a new mechanism. What do tool vendors want to do with this?

Antoine: I use input derivatives up to 2nd order. If the master provides more, it will be considered as an error.

Markus: So the master expects the FMU to use a specific polynomial.

Antoine: we use the definition of the polynomial of FMI 1.0 Appendix D

Markus: then we have 2 mechanism to do the same things

Christian: with IVA the master could do more.  E.g. use exact values (waveform relaxation).

Markus: We are quite far, as IVA is part of BCS. There remain two possibilities: 
1. ) input derivatives are the mathematical derivatives. The FMU decides what to do with them
2.) Input derivatives are coefficients of a specified polynomial that the FMU has to use on the current communication interval.

Christian: 2.) could be also realized with IVA, but one could keep it for compatibility with FMI 2.0.

Markus: ... and for simplicity compared with IVA. 
What are the use cases for 1.)?

Opinions: 
Markus As I do not have a use case for 1), I prefer 2)
Antoine: There are cases where the FMU cannot make use of higher derivatives (e.g. 10th). we have to fix the order. We could have an attribute
Christian: or fix the order in the specification.

Markus: we could have the information that one signal  is the derivative of another signal by providing this information in the model structure.  Thus there is already a mechanism to provide the exact derivative. 

Christian: I agree for defining the interpolation polynomial.

We propose a poll: 
1. ) input derivatives are the mathematical derivatives. The FMU decides what to do with them
2.) Input derivatives are coefficients of a specified polynomial that the FMU has to use on the current communication interval. We use the taylor polynomial similar to FMI 1.0 appendix D. Additionally the FMU has an attribute on the maximum order of input deriviatives, replacing the attribute can interpolate inputs.

**Comment by chrbertsch** [comment]
Superseded by #921
